compiler/hash-intrinsics/src/utils.rs

```diff
@@ -53,6 +76,96 @@ pub trait PrimitiveUtils: AccessToPrimitives + Sized {
     fn new_never_ty(&self) -> TyId {
         self.new_ty(DataTy { args: self.new_empty_args(), data_def: self.primitives().never() })
     }
+
+    /// Get the given type as a literal type if possible.
+    fn try_use_ty_as_lit_ty(&self, ty: TyId) -> Option<LitTy> {
+        match self.get_ty(ty) {
+            Ty::Data(data) => match data.data_def {
+                d if d == self.primitives().i8() => Some(LitTy::I8),
+                d if d == self.primitives().u8() => Some(LitTy::U8),
+                d if d == self.primitives().i16() => Some(LitTy::I16),
+                d if d == self.primitives().u16() => Some(LitTy::U16),
+                d if d == self.primitives().i32() => Some(LitTy::I32),
+                d if d == self.primitives().u32() => Some(LitTy::U32),
+                d if d == self.primitives().i64() => Some(LitTy::I64),
+                d if d == self.primitives().u64() => Some(LitTy::U64),
+                d if d == self.primitives().u128() => Some(LitTy::U128),
+                d if d == self.primitives().i128() => Some(LitTy::I128),
+                d if d == self.primitives().ibig() => Some(LitTy::IBig),
+                d if d == self.primitives().ubig() => Some(LitTy::UBig),
+                d if d == self.primitives().f32() => Some(LitTy::F32),
+                d if d == self.primitives().f64() => Some(LitTy::F64),
+                d if d == self.primitives().bool() => Some(LitTy::Bool),
+                d if d == self.primitives().char() => Some(LitTy::Char),
+                _ => None,
+            },
+            _ => None,
+        }
+    }
+
+    /// Get the given term as a float literal if possible.
+    fn create_term_from_float_lit<L: Into<f64>>(&self, lit: L) -> TermId {
+        self.new_term(Term::Prim(PrimTerm::Lit(Lit::Float(FloatLit {
+            underlying: ast::FloatLit {
+                kind: ast::FloatLitKind::Unsuffixed,
+                value: CONSTANT_MAP.create_f64_float_constant(lit.into(), None),
+            },
+        }))))
+    }
+
+    /// Get the given term as a float literal if possible.
+    fn try_use_term_as_float_lit<L: TryFrom<f64>>(&self, term: TermId) -> Option<L> {
+        match self.get_term(term) {
+            Term::Prim(PrimTerm::Lit(Lit::Float(i))) => i.value().try_into().ok(),
+            _ => None,
+        }
+    }
+
+    /// Get the given term as a float literal if possible.
+    fn create_term_from_integer_lit<L: Into<BigInt>>(&self, lit: L, int_ty: IntTy) -> TermId {
+        self.new_term(Term::Prim(PrimTerm::Lit(Lit::Int(IntLit {
+            underlying: ast::IntLit {
+                kind: ast::IntLitKind::Unsuffixed,
+                value: CONSTANT_MAP.create_int_constant(IntConstant::from_big_int(
+                    lit.into(),
+                    int_ty,
+                    None,
+                )),
```

Comment:
this will have to change into with the new constant stuff since there is no more `ty` on it.